java/core/src/main/java/com/microsoft/bond/BondType.java

```diff
@@ -0,0 +1,514 @@
+     * which is assumed to be already deserialized earlier and whose value is available in the
+     * {@link TaggedDeserializationContext#readFieldResult} field of the passed context argument.
+     * This method is not recommended for fields with Java primitive (non-object) types. To deserialize fields with
+     * primitive values use the static helper method defined in each singleton class for a primitive type.
+     *
+     * @param context contains the runtime context of the deserialization
+     * @param field   descriptor of the field
+     * @return the deserialized value
+     * @throws IOException if an I/O error occurred
+     */
+    protected abstract T deserializeField(
+            TaggedDeserializationContext context,
+            StructBondType.StructField<T> field) throws IOException;
+
+    /**
+     * Deserializes a struct field of this type with "nothing" as the default value from a tagged protocol reader,
+     * excluding field metadata which is assumed to be already deserialized earlier and whose value is available
+     * in the {@link TaggedDeserializationContext#readFieldResult} field of the passed context argument.
+     * This method is not recommended for fields with Java primitive (non-object) types. To deserialize fields with
+     * primitive values use the static helper method defined in each singleton class for a primitive type.
+     *
+     * @param context contains the runtime context of the deserialization
+     * @param field   descriptor of the field
+     * @return the deserialized value
+     * @throws IOException if an I/O error occurred
+     */
+    protected final SomethingObject<T> deserializeSomethingField(
+            TaggedDeserializationContext context,
+            StructBondType.StructField<T> field) throws IOException {
+        return Something.wrap(this.deserializeField(context, field));
+    }
+
+    @Override
+    public final int hashCode() {
+        return this.precomputedHashCode;
+    }
+
+    @Override
+    public final boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        } else if (obj == null) {
+            return false;
+        } else if (this.getClass() != obj.getClass()) {
+            return false;
+        } else {
+            BondType<?> that = (BondType<?>) obj;
+            if (this.precomputedHashCode != that.precomputedHashCode) {
+                return false;
+            } else {
+                return this.equalsInternal(that);
+            }
+        }
+    }
+
+    // helper for subclasses to compute hash code
+    static int multiplyAndShift(int value, int n) {
+        return (value * n) ^ (value >>> (32 - n));
+    }
+
+    // implemented by subclasses to test type decriptors for equality
+    // (the argument is not null and can be safely cast to the actual subclass)
+    abstract boolean equalsInternal(BondType<?> obj);
+
+    @Override
+    public final String toString() {
+        return this.getFullName();
```

Comment:
I understand, but it doesn't contain anything that would point a reader back to the `BondType` hierarchy. I would like it to do so.